healthcheck/tor_connection.go

```diff
@@ -0,0 +1,85 @@
+package healthcheck
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"syscall"
+
+	"github.com/lightningnetwork/lnd/tor"
+)
+
+// CheckTorServiceStatus checks the Onion Service's info to decide whether the
+// Tor daemon is up and running. We will get an EOF or a broken pipe error if
+// the Tor daemon is stopped/restarted as the previously created socket
+// connection is no longer open. In this case, we will attempt a restart on our
+// tor controller. If the tor daemon comes back, a new socket connection will
+// then be created.
+func CheckTorServiceStatus(tc *tor.Controller, createService func() error) error {
+	err := tc.GetServiceInfo()
+	if err == nil {
+		return nil
+	}
+
+	log.Debugf("Checking tor service got: %v", err)
+
+	switch {
+	// We will get an EOF if the connection is lost. In this case, we will
+	// return an error and wait for the Tor daemon to come back. We won't
+	// attempt to make a new connection since we know Tor daemon is down.
+	case errors.Is(err, io.EOF), errors.Is(err, syscall.ECONNREFUSED):
+		log.Warnf("Tor connection lost, attempting a tor controller " +
+			"re-connection in next check")
+
+		return fmt.Errorf("Tor daemon connection lost, " +
+			"check if Tor is up and running")
+
+	// We will get a broken pipe error when we use the old connection to
```

Comment:
Old connection here is the existing connection we have before the tor daemon restarted (in this scenario)?

Reply:
Yes! Revised the docs to make it more clear.